quinn/src/endpoint.rs

```diff
@@ -558,3 +577,32 @@ where
         &self.0
     }
 }
+
+#[cfg(target_os = "linux")]
+/// Expand GSO-layout Transmits into single segment Transmits
+fn flatten_gso(transmits: &mut VecDeque<proto::Transmit>) {
```

Comment:
Not sure whether it's worth it. The first transmit was already lost at the point, and the connection doesn't know. Before it sets something up for retransmits the capability change will also have propagated, and there is no need for flatten for the retry.

Instead of copying contents here we could also have a cheaper fallback which just does sends multiple messages and just references the payload slices. But that might obviously be even more invasive.

Reply:
The latter was my first idea, but it turns out to be a huge pain because sendmmsg can have partial success whereas GSO can't(?). With this code, we never discard the failed transmit, so it's not lost, though I agree it's kinda ugly. Are you thinking we might as well just pretend we lost the packets in this case and deal with the transient extra latency? We do need to handle the possibility of multiple bad flights, since there could be multiple connections in flight and getting new data to send vs. updating the capabilities is racey.

Reply:
> The latter was my first idea, but it turns out to be a huge pain because sendmmsg can have partial success whereas GSO can't(?).

I think that's the theory, but I'm not sure whether it matters in practice. It looks to me like all the `sendmsg` APIs will signal success anyway, and packets will get dropped later on a lower layer (or they might even get dropped when success is signaled).

> Are you thinking we might as well just pretend we lost the packets in this case and deal with the transient extra latency? 

Yes, that was the idea. But I kind of agree its ugly that those systems will then encounter an additional handshake delay - which makes QUIC certainly perform less than TCP + TLS1.2.  Some better understanding on how widespread those broken systems are would be interesting to determine whether that's a reasonable path forward